src/debug.h

```diff
@@ -197,6 +197,50 @@ void limitDebugClass( int );
  */
 bool debug_has_error_been_observed();
 
+/**
+ * Struct for return capturing result
+ */
+template<typename T>
+struct capture_ret {
+    T result;
+    std::string dmsg;
+    bool error_observed;
+};
+
+// Don't use this, use CAPTURE_DEBUGMSG instead.
+void captureStart();
+// Don't use this, use CAPTURE_DEBUGMSG instead.
+std::string captureEnd();
+
+/**
+ * Capturing debugmsg for a member function
+ *  @return capture_ret<R> {result:R, dmsg:string, error_observed: bool }
+ */
+template<typename R, typename O, typename M, typename... Args>
+capture_ret<R> CAPTURE_DEBUGMSG( O &object, M member_func, Args... args )
+{
+    capture_ret<R> ret;
+    captureStart();
+    ret.result = ( object.*member_func )( args... );
+    ret.dmsg = captureEnd();
+    ret.error_observed = !ret.dmsg.empty();
+    return ret;
+}
+/**
+ * Capturing debugmsg for a non-member or static member function
+ *  @return capture_ret<R> {result:R, dmsg:string, error_observed: bool }
+ */
+template<typename R, typename F, typename... Args>
+capture_ret<R> CAPTURE_DEBUGMSG_S( F func, Args... args )
+{
+    capture_ret<R> ret;
+    captureStart();
+    ret.result = func( args... );
+    ret.dmsg = captureEnd();
+    ret.error_observed = !ret.dmsg.empty();
+    return ret;
+}
+
```

Comment:
Since these are only supposed to be used in the tests, you can add a header and a cpp file to the test code that contains `captureStart`, `captureEnd`, but not `CAPTURE_DEBUGMSG` or `CAPTURE_DEBUGMSG_S`. Instead return `dmsg` and `error_observed` directly from `captureEnd`, so you can execute whatever code before calling `captureEnd` instead of being limited to a member or non-member function call and a single templated return value. To access `capturing` and `captured` in both cpp files, you can define them in `debug.cpp` without the `static` keyword, and declare them in the new cpp file using the `extern` keyword (`extern bool capturing;`)

You can also replace `captureStart` and `captureEnd` with a `captureDebugmsg` object that calls `captureStart` on initialization and calls `captureEnd` when explicitly asked to return the captured debug message and when the object is deconstructed. This ensures that `captureEnd` is always called appropriately without creating a function like `CAPTURE_DEBUGMSG`.